Add #1206: Do not require a "friendly name"

Computroniks opened the pull request:
# Description

When creating or updating a monitor, a friendly name will no longer be required for specific monitor types. If it is not specified it will be set either the hostname or URL. In cases where both the URL and hostname are set (possible in instances where monitor type has been changed), the hostname has priority over the URL and the friendly name will be set to the hostname.

The array `friendlyNameRequiredOptions` contains all monitor types that still require a friendly name because they don't have either a URL or hostname field. For any monitor type specified in this array, the friendly name field will still be required. 

Monitor types that no longer require a friendly name:
- HTTP(s)
- TCP Port
- Ping
- HTTP(s) - Keyword
- DNS
- Steam Game Server

Closes #1206 

## Type of change

- User interface (UI)
- New feature (non-breaking change which adds functionality)

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I ran ESLint and other linters for modified files
- [x] I have performed a self-review of my own code and tested it
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings

Computroniks commented:
When I save the friendly name, it can either be `""` or `null`. By using `""` it would prevent issues with this line https://github.com/louislam/uptime-kuma/blob/c811c1ccdea1e65a864d885f8c960520bd1636d5/src/components/MonitorList.vue#L94
but it would also show as an empty string in the prometheus metrics. For some reason it doesn't feel quite right just to leave it as an empty string in the metrics. Theoretically the empty string could just be changed to null when returning the prometheus metrics but this seems a bit messy. Any thoughts on what would be best here?

karelkryda commented:
> When I save the friendly name, it can either be `""` or `null`. By using `""` it would prevent issues with this line
> 
> https://github.com/louislam/uptime-kuma/blob/c811c1ccdea1e65a864d885f8c960520bd1636d5/src/components/MonitorList.vue#L94
> 
> 
> but it would also show as an empty string in the prometheus metrics. For some reason it doesn't feel quite right just to leave it as an empty string in the metrics. Theoretically the empty string could just be changed to null when returning the prometheus metrics but this seems a bit messy. Any thoughts on what would be best here?

@Computroniks, If the monitor name is empty, simply set the monitor name to its URL. So the only thing you need to change is the way the data is stored in the database.

Ie. if the monitor name is empty => `monitor.name = monitor.url` ... this is the best solution in my opinion.

Computroniks commented:
@karelkryda I did consider that here:
> Another way that this feature could be implemented is when a monitor is created, the url is copied into the name field either client or server side. The issue I see with this is if the URL were to change, it could get messy trying to work out if the friendly name should change as well.

https://github.com/louislam/uptime-kuma/issues/1206#issuecomment-1022455220

Whilst it doesn't seem like such of a big deal if someone changes the url and then has to change the friendly name as well, it seems a bit like only half a job to just copy it over to the name field.

karelkryda commented:
@Computroniks I don't see a problem with that.
As I say, if the monitor name and url match when the monitor is edited, reset the monitor name (set the name to empty), and if the monitor name is not entered during editing, set the monitor name to match the monitor url when saving.

Computroniks commented:
Ahh yes, of course. If I just implement a check to see if the URL and friendly name match when a user edits the monitor, I can then work out if the name should change as well. Not sure why I didn't think of that before. Thanks

karelkryda commented:
No problem, I was happy to help 🙂

karelkryda commented on src/pages/EditMonitor.vue:
```diff
@@ -532,6 +551,19 @@ export default {
+            if (this.friendlyNameRequiredOptions.includes(this.monitor.type)) {
+                return true;
+            } else {
+                return null;
```
It would be better to simply return false here

Computroniks replied:
Fixed

karelkryda commented on src/pages/EditMonitor.vue:
```diff
@@ -317,6 +317,7 @@ export default {
             },
             acceptedStatusCodeOptions: [],
             dnsresolvetypeOptions: [],
+            friendlyNameRequiredOptions: [],
```
You can enter options here directly because the array is not populated dynamically

Computroniks replied:
I put it with the other two in `init()` as it helps to keep it consistent with them. `dnsresolvetypeOptions` isn't dynamic either so it kind of makes sense to keep them consistent.

karelkryda replied:
ok, no problem

karelkryda commented on src/pages/EditMonitor.vue:
```diff
@@ -413,12 +414,17 @@ export default {
 
         this.acceptedStatusCodeOptions = acceptedStatusCodeOptions;
         this.dnsresolvetypeOptions = dnsresolvetypeOptions;
+        this.friendlyNameRequiredOptions = friendlyNameRequiredOptions;
```
This is unnecessary, see. comment above

karelkryda commented on src/pages/EditMonitor.vue:
```diff
@@ -413,12 +414,17 @@ export default {
             "TXT",
         ];
 
+        let friendlyNameRequiredOptions = [
+            "push",
+        ];
+
```
This is unnecessary, see. comment above

karelkryda commented on src/pages/EditMonitor.vue:
```diff
@@ -486,6 +495,16 @@ export default {
+            // Check if friendly name has been supplied. If not, use URL
+            // or hostname
+            if (this.monitor.name === "") {
+                if (this.monitor.hostname === null || this.monitor.hostname === "") {
+                    this.monitor.name = this.monitor.url;
+                } else {
+                    this.monitor.name = this.monitor.hostname;
+                }
```
You can use `this.monitor.name = this.monitor.hostname || this.monitor.url;`

Computroniks replied:
Fixed

karelkryda reviewed (changes requested):
Please fix a few things 🙂

karelkryda commented on src/pages/EditMonitor.vue:
```diff
@@ -532,6 +545,19 @@ export default {
+         * @returns {boolean}
+         */
+        friendlyNameRequired() {
+            if (this.friendlyNameRequiredOptions.includes(this.monitor.type)) {
+                return true;
+            } else {
+                return false;
+            }
```
oh also here you can do just

`return this.friendlyNameRequiredOptions.includes(this.monitor.type);`

Computroniks replied:
Fixed. Thanks for your help 👍

Saibamen commented on src/pages/EditMonitor.vue:
```diff
@@ -450,7 +456,10 @@ export default {
                     if (res.ok) {
                         this.monitor = res.monitor;
-
+                        // Handling for when friendly name isn't set
+                        if (this.monitor.name === this.monitor.url || this.monitor.name === this.monitor.hostname) {
+                            this.monitor.name = "";
+                        }
                         // Handling for monitors that are created before 1.7.0
```
```suggestion

                        // Handling for when friendly name isn't set
                        if (this.monitor.name === this.monitor.url || this.monitor.name === this.monitor.hostname) {
                            this.monitor.name = "";
                        }
                        
                        // Handling for monitors that are created before 1.7.0
```

Computroniks commented:
Anything I can do to help get this merged?

gaby commented:
@louislam Can we get this merge?

Computroniks commented:
Fixed merge conflicts

Computroniks commented:
Well, I don't know why the checks are failing, the error indicates a line that I didn't even change.
```
✓ 77 modules transformed.
[rollup-plugin-dynamic-import-variables] Unexpected token (294:8)
file: /home/runner/work/uptime-kuma/uptime-kuma/src/pages/EditMonitor.vue:294:8
error during build:
SyntaxError: Unexpected token (294:8)
    at Parser.pp$4.raise (/home/runner/work/uptime-kuma/uptime-kuma/node_modules/rollup/dist/shared/rollup.js:19636:13)
    at Parser.pp$9.unexpected (/home/runner/work/uptime-kuma/uptime-kuma/node_modules/rollup/dist/shared/rollup.js:16932:8)
    at Parser.pp$9.expect (/home/runner/work/uptime-kuma/uptime-kuma/node_modules/rollup/dist/shared/rollup.js:16926:26)
    at Parser.pp$5.parseObj (/home/runner/work/uptime-kuma/uptime-kuma/node_modules/rollup/dist/shared/rollup.js:19248:[12](https://github.com/louislam/uptime-kuma/runs/5835807857?check_suite_focus=true#step:5:12))
    at Parser.pp$5.parseExprAtom (/home/runner/work/uptime-kuma/uptime-kuma/node_modules/rollup/dist/shared/rollup.js:18987:[17](https://github.com/louislam/uptime-kuma/runs/5835807857?check_suite_focus=true#step:5:17))
    at Parser.pp$5.parseExprSubscripts (/home/runner/work/uptime-kuma/uptime-kuma/node_modules/rollup/dist/shared/rollup.js:[18](https://github.com/louislam/uptime-kuma/runs/5835807857?check_suite_focus=true#step:5:18)803:[19](https://github.com/louislam/uptime-kuma/runs/5835807857?check_suite_focus=true#step:5:19))
    at Parser.pp$5.parseMaybeUnary (/home/runner/work/uptime-kuma/uptime-kuma/node_modules/rollup/dist/shared/rollup.js:18769:17)
    at Parser.pp$5.parseExprOps (/home/runner/work/uptime-kuma/uptime-kuma/node_modules/rollup/dist/shared/rollup.js:18696:19)
    at Parser.pp$5.parseMaybeConditional (/home/runner/work/uptime-kuma/uptime-kuma/node_modules/rollup/dist/shared/rollup.js:18679:19)
    at Parser.pp$5.parseMaybeAssign (/home/runner/work/uptime-kuma/uptime-kuma/node_modules/rollup/dist/shared/rollup.js:18646:19)
Error: Process completed with exit code 1.
```

Computroniks commented:
Anything I can do tp help get this merged?

Computroniks commented:
Fixed merge conflicts

Computroniks commented:
Is there anything stopping this PR from being merged?

louislam commented:
Just testing it with different monitor types, found out that the friendly name sometimes do not follow the current url.

![image](https://user-images.githubusercontent.com/1336778/213697825-cc7b8cf6-e28e-4837-8fbb-2635dc3131e6.png)

This video may help.
https://user-images.githubusercontent.com/1336778/213697855-7e083688-5b61-4d79-9114-a26c6745d50c.mp4

Computroniks commented:
This is probably because the docker monitor was added after this PR was first created and I haven't updated it to fix. I will get on that

Computroniks commented:
This PR should have now been updated to match all of the monitor types added recently. I have also modified the logic for selecting the field to use for the friendly name so that we don't end up with the issue where we are using a http monitor but the tcp hostname is used. 

I also added a custom case for TCP monitors to include the port number being monitored.

Computroniks commented:
That should be the merge conflicts now fixed

Saibamen commented on src/pages/EditMonitor.vue:
```diff
@@ -1243,6 +1254,16 @@ message HealthCheckResponse {
                         this.monitor = res.monitor;
 
+                        // Handling for when friendly name isn't set
+                        if (
+                            this.monitor.name === this.monitor.url
```
```suggestion
                        if (this.monitor.name === this.monitor.url
```

Computroniks replied:
Personally I prefer it on a new line because it fits better with how we would write function arguments or arrays that split over new lines.

Saibamen reviewed (commented):
Looks OK

karelkryda reviewed (approved):
LGTM :)

Computroniks commented:
@louislam This has been open for 2 years, are there any plans of merging it?

CommanderStorm commented:
I merged https://github.com/louislam/uptime-kuma/pull/5795 instead => a similar feature now exists
